File: backend/internal/model/activity.go
package model

import "time"

// ActivityStatus 活动状态
type ActivityStatus string

const (
	ActivityStatusDraft    ActivityStatus = "draft"    // 草稿
	ActivityStatusPreheat  ActivityStatus = "preheat"  // 预热中（可见但不可参与）
	ActivityStatusActive   ActivityStatus = "active"   // 进行中
	ActivityStatusPaused   ActivityStatus = "paused"   // 已暂停
	ActivityStatusEnded    ActivityStatus = "ended"    // 已结束
	ActivityStatusCanceled ActivityStatus = "canceled" // 已取消
)

// ActivityType 活动类型
type ActivityType string

const (
	ActivityTypeCoupon   ActivityType = "coupon"   // 优惠券发放活动
	ActivityTypeDiscount ActivityType = "discount" // 限时折扣活动（预留）
	ActivityTypeGift     ActivityType = "gift"     // 赠品活动（预留）
)

// Activity 活动
type Activity struct {
	Base
	Name        string         `json:"name" gorm:"size:128;not null"`                         // 活动名称
	Description string         `json:"description" gorm:"type:text"`                          // 活动描述
	Type        ActivityType   `json:"type" gorm:"size:32;not null;index"`                    // 活动类型
	Status      ActivityStatus `json:"status" gorm:"size:32;default:'draft';index"`           // 活动状态
	CoverURL    string         `json:"coverUrl,omitempty" gorm:"column:cover_url;size:512"`   // 活动封面图
	BannerURL   string         `json:"bannerUrl,omitempty" gorm:"column:banner_url;size:512"` // 活动Banner图

	// 时间控制
	PreheatAt *time.Time `json:"preheatAt,omitempty" gorm:"column:preheat_at"` // 预热开始时间
	StartAt   time.Time  `json:"startAt" gorm:"column:start_at;not null"`      // 活动开始时间
	EndAt     time.Time  `json:"endAt" gorm:"column:end_at;not null"`          // 活动结束时间

	// 参与限制
	TotalLimit   int `json:"totalLimit" gorm:"column:total_limit;default:0"`      // 总参与次数限制（0=无限制）
	DailyLimit   int `json:"dailyLimit" gorm:"column:daily_limit;default:0"`      // 每日参与次数限制（0=无限制）
	PerUserLimit int `json:"perUserLimit" gorm:"column:per_user_limit;default:1"` // 每人参与次数限制

	// 统计
	TotalParticipants int `json:"totalParticipants" gorm:"column:total_participants;default:0"` // 总参与人数
	TodayParticipants int `json:"todayParticipants" gorm:"column:today_participants;default:0"` // 今日参与人数
	TotalClaimed      int `json:"totalClaimed" gorm:"column:total_claimed;default:0"`           // 总领取次数

	// 配置
	AllowVipStack bool   `json:"allowVipStack" gorm:"column:allow_vip_stack;default:false"` // 是否允许与VIP折扣叠加
	Rules         string `json:"rules,omitempty" gorm:"type:text"`                          // 活动规则说明
	SortOrder     int    `json:"sortOrder" gorm:"column:sort_order;default:0"`              // 排序（越小越靠前）
	IsVisible     bool   `json:"isVisible" gorm:"column:is_visible;default:true"`           // 是否在前端展示

	// 关联
	Rewards []ActivityReward `json:"rewards,omitempty" gorm:"foreignKey:ActivityID"`
}

// TableName 指定表名
func (Activity) TableName() string {
	return "activities"
}

// IsInPreheat 是否在预热期
func (a *Activity) IsInPreheat() bool {
	now := time.Now()
	if a.PreheatAt == nil {
		return false
	}
	return now.After(*a.PreheatAt) && now.Before(a.StartAt)
}

// IsActive 是否在活动期间
func (a *Activity) IsActive() bool {
	now := time.Now()
	return a.Status == ActivityStatusActive && now.After(a.StartAt) && now.Before(a.EndAt)
}

// ActivityReward 活动奖励配置
type ActivityReward struct {
	Base
	ActivityID       uint64 `json:"activityId" gorm:"column:activity_id;not null;index"`        // 活动ID
	CouponTemplateID uint64 `json:"couponTemplateId" gorm:"column:coupon_template_id;not null"` // 优惠券模板ID
	CouponCount      int    `json:"couponCount" gorm:"column:coupon_count;default:1"`           // 每次发放数量
	Probability      int    `json:"probability" gorm:"default:100"`                             // 发放概率（1-100，预留抽奖用）
	TotalStock       int    `json:"totalStock" gorm:"column:total_stock;default:0"`             // 总库存（0=无限制）
	RemainingStock   int    `json:"remainingStock" gorm:"column:remaining_stock;default:0"`     // 剩余库存
	SortOrder        int    `json:"sortOrder" gorm:"column:sort_order;default:0"`               // 排序

	// Relations
	Activity       *Activity       `json:"activity,omitempty" gorm:"foreignKey:ActivityID"`
	CouponTemplate *CouponTemplate `json:"couponTemplate,omitempty" gorm:"foreignKey:CouponTemplateID"`
}

// TableName 指定表名
func (ActivityReward) TableName() string {
	return "activity_rewards"
}

// ActivityParticipation 活动参与记录
type ActivityParticipation struct {
	Base
	ActivityID uint64    `json:"activityId" gorm:"column:activity_id;not null;index"`   // 活动ID
	UserID     uint64    `json:"userId" gorm:"column:user_id;not null;index"`           // 用户ID
	RewardID   uint64    `json:"rewardId" gorm:"column:reward_id;not null"`             // 奖励配置ID
	CouponIDs  string    `json:"couponIds,omitempty" gorm:"column:coupon_ids;size:512"` // 发放的优惠券ID列表（JSON数组）
	ClaimedAt  time.Time `json:"claimedAt" gorm:"column:claimed_at;not null"`           // 领取时间
	ClientIP   string    `json:"clientIp,omitempty" gorm:"column:client_ip;size:64"`    // 客户端IP

	// Relations
	Activity *Activity       `json:"activity,omitempty" gorm:"foreignKey:ActivityID"`
	User     *User           `json:"user,omitempty" gorm:"foreignKey:UserID"`
	Reward   *ActivityReward `json:"reward,omitempty" gorm:"foreignKey:RewardID"`
}

// TableName 指定表名
func (ActivityParticipation) TableName() string {
	return "activity_participations"
}

// ActivityDailyStats 活动每日统计（用于每日限制重置）
type ActivityDailyStats struct {
	Base
	ActivityID   uint64    `json:"activityId" gorm:"column:activity_id;not null;uniqueIndex:idx_activity_date"` // 活动ID
	StatsDate    time.Time `json:"statsDate" gorm:"column:stats_date;not null;uniqueIndex:idx_activity_date"`   // 统计日期
	Participants int       `json:"participants" gorm:"default:0"`                                               // 当日参与人数
	ClaimCount   int       `json:"claimCount" gorm:"column:claim_count;default:0"`                              // 当日领取次数
}

// TableName 指定表名
func (ActivityDailyStats) TableName() string {
	return "activity_daily_stats"
}

File: admin/src/api/activity.ts
import apiClient from './client';
import type { ApiResponse, Pagination } from '@/types/api';

// Re-export for backward compatibility
export type { ApiResponse };

// ============================================================================
// Type Definitions
// ============================================================================

/**
 * Activity Status (活动状态)
 */
export type ActivityStatus = 'draft' | 'preheat' | 'active' | 'paused' | 'ended' | 'canceled';

/**
 * Activity Type (活动类型)
 */
export type ActivityType = 'coupon' | 'discount' | 'gift';

/**
 * Activity Interface
 * Matches backend model.Activity
 */
export interface Activity {
    id: number;
    name: string;
    description?: string;
    type: ActivityType;
    status: ActivityStatus;
    coverUrl?: string;
    bannerUrl?: string;

    // Time control
    preheatAt?: string;
    startAt: string;
    endAt: string;

    // Participation limits
    totalLimit: number;
    dailyLimit: number;
    perUserLimit: number;

    // Statistics
    totalParticipants: number;
    todayParticipants: number;
    totalClaimed: number;

    // Configuration
    allowVipStack: boolean;
    rules?: string;
    sortOrder: number;
    isVisible: boolean;

    // Relations
    rewards?: ActivityReward[];

    // Timestamps
    createdAt: string;
    updatedAt: string;
}

/**
 * Activity Reward Interface
 * Matches backend model.ActivityReward
 */
export interface ActivityReward {
    id: number;
    activityId: number;
    couponTemplateId: number;
    couponCount: number;
    probability: number;
    totalStock: number;
    remainingStock: number;
    sortOrder: number;

    // Relations
    activity?: Activity;
    couponTemplate?: {
        id: number;
        name: string;
        type: string;
    };

    // Timestamps
    createdAt: string;
    updatedAt: string;
}

/**
 * Activity Participation Interface
 * Matches backend model.ActivityParticipation
 */
export interface ActivityParticipation {
    id: number;
    activityId: number;
    userId: number;
    rewardId: number;
    couponIds?: string;
    claimedAt: string;
    clientIp?: string;

    // Relations
    activity?: Activity;
    user?: {
        id: number;
        name: string;
        avatarUrl?: string;
    };
    reward?: ActivityReward;

    // Timestamps
    createdAt: string;
    updatedAt: string;
}

/**
 * Activity Statistics
 */
export interface ActivityStats {
    activityId: number;
    totalParticipants: number;
    todayParticipants: number;
    totalClaimed: number;
    remainingStock?: number;
}

/**
 * All Activities Statistics Overview
 */
export interface AllActivityStats {
    totalActivities: number;
    activeActivities: number;
    draftActivities: number;
    endedActivities: number;
    totalParticipants: number;
    totalClaimed: number;
}

/**
 * Batch Operation Result
 */
export interface BatchOperationResult {
    successCount: number;
    failedCount: number;
    failedIds: number[];
    errors: string[];
}

// ============================================================================
// Request DTOs
// ============================================================================

/**
 * Create Activity Request
 */
export interface CreateActivityDto {
    name: string;
    description?: string;
    type: ActivityType;
    coverUrl?: string;
    bannerUrl?: string;
    preheatAt?: string;
    startAt: string;
    endAt: string;
    totalLimit?: number;
    dailyLimit?: number;
    perUserLimit?: number;
    allowVipStack?: boolean;
    rules?: string;
    sortOrder?: number;
    isVisible?: boolean;
    status?: ActivityStatus;
}

/**
 * Update Activity Request
 */
export type UpdateActivityDto = Partial<CreateActivityDto>;

/**
 * Activity Query Parameters
 */
export interface ActivityQueryParams {
    page?: number;
    page_size?: number;
    keyword?: string;
    type?: ActivityType;
    status?: ActivityStatus;
    isVisible?: boolean;
    startTime?: string;
    endTime?: string;
}

/**
 * Update Activity Status Request
 */
export interface UpdateActivityStatusDto {
    status: ActivityStatus;
}

/**
 * Create Reward Request
 */
export interface CreateRewardDto {
    activityId: number;
    couponTemplateId: number;
    couponCount: number;
    probability: number;
    totalStock?: number;
    sortOrder?: number;
}

/**
 * Update Reward Request
 */
export type UpdateRewardDto = Partial<CreateRewardDto>;

/**
 * Participation Query Parameters
 */
export interface ParticipationQueryParams {
    page?: number;
    page_size?: number;
    activityId?: number;
    userId?: number;
    startTime?: string;
    endTime?: string;
}

/**
 * Batch Delete Activities Request
 */
export interface BatchDeleteActivitiesDto {
    ids: number[];
}

/**
 * Batch Update Activities Status Request
 */
export interface BatchUpdateActivitiesStatusDto {
    ids: number[];
    status: ActivityStatus;
}

/**
 * Batch Publish Activities Request
 */
export interface BatchPublishActivitiesDto {
    ids: number[];
    isVisible: boolean;
}

// ============================================================================
// API Client
// ============================================================================

/**
 * Activity API
 * Provides methods for managing activities, rewards, and participations
 */
export const activityApi = {
    // ========================================================================
    // Activity Management
    // ========================================================================

    /**
     * Get activities list
     * GET /admin/activities
     */
    getActivities: (params?: ActivityQueryParams) =>
        apiClient.get<ApiResponse<Activity[]>>('/admin/activities', { params }),

    /**
     * Get activity detail
     * GET /admin/activities/:id
     */
    getActivityDetail: (id: number) =>
        apiClient.get<ApiResponse<Activity>>(`/admin/activities/${id}`),

    /**
     * Create new activity
     * POST /admin/activities
     */
    createActivity: (data: CreateActivityDto) =>
        apiClient.post<ApiResponse<Activity>>('/admin/activities', data),

    /**
     * Update activity
     * PUT /admin/activities/:id
     */
    updateActivity: (id: number, data: UpdateActivityDto) =>
        apiClient.put<ApiResponse<Activity>>(`/admin/activities/${id}`, data),

    /**
     * Delete activity
     * DELETE /admin/activities/:id
     */
    deleteActivity: (id: number) =>
        apiClient.delete<ApiResponse<void>>(`/admin/activities/${id}`),

    /**
     * Update activity status
     * PUT /admin/activities/:id/status
     */
    updateActivityStatus: (id: number, data: UpdateActivityStatusDto) =>
        apiClient.put<ApiResponse<{ message: string }>>(`/admin/activities/${id}/status`, data),

    /**
     * Publish activity (convenience method)
     * Updates status to 'active' and isVisible to true
     */
    publishActivity: (id: number) =>
        activityApi.updateActivity(id, { status: 'active', isVisible: true }),

    /**
     * Unpublish activity (convenience method)
     * Sets isVisible to false
     */
    unpublishActivity: (id: number) =>
        activityApi.updateActivity(id, { isVisible: false }),

    // ========================================================================
    // Batch Operations
    // ========================================================================

    /**
     * Batch delete activities
     * DELETE /admin/activities/batch
     */
    batchDeleteActivities: (data: BatchDeleteActivitiesDto) =>
        apiClient.delete<ApiResponse<BatchOperationResult>>('/admin/activities/batch', { data }),

    /**
     * Batch update activities status
     * PUT /admin/activities/batch/status
     */
    batchUpdateActivitiesStatus: (data: BatchUpdateActivitiesStatusDto) =>
        apiClient.put<ApiResponse<BatchOperationResult>>('/admin/activities/batch/status', data),

    /**
     * Batch publish/unpublish activities
     * PUT /admin/activities/batch/publish
     */
    batchPublishActivities: (data: BatchPublishActivitiesDto) =>
        apiClient.put<ApiResponse<BatchOperationResult>>('/admin/activities/batch/publish', data),

    // ========================================================================
    // Reward Management
    // ========================================================================

    /**
     * Get rewards by activity ID
     * GET /admin/activities/:id/rewards
     */
    getActivityRewards: (activityId: number) =>
        apiClient.get<ApiResponse<ActivityReward[]>>(`/admin/activities/${activityId}/rewards`),

    /**
     * Get reward detail
     * GET /admin/activities/:id/rewards/:rewardId
     */
    getRewardDetail: (activityId: number, rewardId: number) =>
        apiClient.get<ApiResponse<ActivityReward>>(`/admin/activities/${activityId}/rewards/${rewardId}`),

    /**
     * Create new reward
     * POST /admin/activities/rewards
     */
    createReward: (data: CreateRewardDto) =>
        apiClient.post<ApiResponse<ActivityReward>>('/admin/activities/rewards', data),

    /**
     * Update reward
     * PUT /admin/activities/rewards/:rewardId
     */
    updateReward: (rewardId: number, data: UpdateRewardDto) =>
        apiClient.put<ApiResponse<ActivityReward>>(`/admin/activities/rewards/${rewardId}`, data),

    /**
     * Delete reward
     * DELETE /admin/activities/rewards/:rewardId
     */
    deleteReward: (rewardId: number) =>
        apiClient.delete<ApiResponse<void>>(`/admin/activities/rewards/${rewardId}`),

    // ========================================================================
    // Participation Records
    // ========================================================================

    /**
     * Get participation records list
     * GET /admin/activities/participations
     */
    getParticipations: (params?: ParticipationQueryParams) =>
        apiClient.get<ApiResponse<ActivityParticipation[]>>('/admin/activities/participations', { params }),

    // ========================================================================
    // Statistics
    // ========================================================================

    /**
     * Get activity statistics
     * GET /admin/activities/:id/stats
     */
    getActivityStats: (id: number) =>
        apiClient.get<ApiResponse<ActivityStats>>(`/admin/activities/${id}/stats`),

    /**
     * Get all activities statistics overview
     * GET /admin/activities/stats
     */
    getAllActivityStats: () =>
        apiClient.get<ApiResponse<AllActivityStats>>('/admin/activities/stats'),
};

// ============================================================================
// Helper Functions
// ============================================================================

/**
 * Format activity type display name
 */
export const getActivityTypeLabel = (type: ActivityType): string => {
    const labels: Record<ActivityType, string> = {
        coupon: '优惠券发放',
        discount: '限时折扣',
        gift: '赠品活动',
    };
    return labels[type] || type;
};

/**
 * Format activity status display name
 */
export const getActivityStatusLabel = (status: ActivityStatus): string => {
    const labels: Record<ActivityStatus, string> = {
        draft: '草稿',
        preheat: '预热中',
        active: '进行中',
        paused: '已暂停',
        ended: '已结束',
        canceled: '已取消',
    };
    return labels[status] || status;
};

/**
 * Format activity status color (for Ant Design tags)
 */
export const getActivityStatusColor = (status: ActivityStatus): string => {
    const colors: Record<ActivityStatus, string> = {
        draft: 'default',
        preheat: 'blue',
        active: 'green',
        paused: 'orange',
        ended: 'default',
        canceled: 'red',
    };
    return colors[status] || 'default';
};

/**
 * Check if activity is currently active
 */
export const isActivityActive = (activity: Activity): boolean => {
    const now = new Date();
    const startAt = new Date(activity.startAt);
    const endAt = new Date(activity.endAt);
    return activity.status === 'active' && now >= startAt && now <= endAt;
};

/**
 * Check if activity is in preheat period
 */
export const isActivityInPreheat = (activity: Activity): boolean => {
    if (!activity.preheatAt) return false;
    const now = new Date();
    const preheatAt = new Date(activity.preheatAt);
    const startAt = new Date(activity.startAt);
    return now >= preheatAt && now < startAt;
};

/**
 * Check if activity has ended
 */
export const isActivityEnded = (activity: Activity): boolean => {
    const now = new Date();
    const endAt = new Date(activity.endAt);
    return now > endAt;
};

/**
 * Calculate stock percentage
 */
export const calculateStockPercentage = (reward: ActivityReward): number => {
    if (reward.totalStock === 0) return 100; // Unlimited stock
    if (reward.totalStock <= 0) return 0;
    return Math.round((reward.remainingStock / reward.totalStock) * 100);
};

/**
 * Format participant count
 */
export const formatParticipantCount = (count: number): string => {
    if (count >= 10000) {
        return `${(count / 10000).toFixed(1)}万`;
    }
    if (count >= 1000) {
        return `${(count / 1000).toFixed(1)}k`;
    }
    return count.toString();
};

/**
 * Parse JSON array string to array
 */
export const parseCouponIds = (jsonString: string): number[] => {
    try {
        const parsed = JSON.parse(jsonString || '[]');
        return Array.isArray(parsed) ? parsed : [];
    } catch {
        return [];
    }
};

/**
 * Get activity time range display text
 */
export const getActivityTimeRange = (activity: Activity): string => {
    const start = new Date(activity.startAt);
    const end = new Date(activity.endAt);
    const format = (date: Date) => {
        return `${date.getMonth() + 1}/${date.getDate()}`;
    };
    return `${format(start)} - ${format(end)}`;
};

/**
 * Check if activity can be edited
 */
export const canEditActivity = (activity: Activity): boolean => {
    return activity.status === 'draft' || activity.status === 'preheat';
};

/**
 * Check if activity can be deleted
 */
export const canDeleteActivity = (activity: Activity): boolean => {
    return activity.status === 'draft' || activity.status === 'canceled' || activity.status === 'ended';
};

// ============================================================================
// Default export
// ============================================================================

export default activityApi;
